### PR TITLE
Enable setting multipart to true/false by defining it as a custom client context parameter

### PR DESCRIPTION
**Pull request by anirudh9391**
Multipart upload can be set as a client configuration parameter and is accessible at the time of instantiation. This allows classes that instantiate the S3AsyncClient to validate this parameter. 

## Modifications
1)Incorporated multiPartEnabled as a custom client configuration parameter by defining it in customization.config. The codegen module generates getter and setter methods for this parameter. 
2) By default, set multipartEnabled = true in the S3AsyncClient
3) Removed code that sets multiPart Enabled on the client directly

## Testing
Unit Tests to validate that 
1) Disabling multiPart code path logs warnings
2) Default S3AsyncClient has multiPartEnabled = true by default. 
3) Setting MultiPartEnabled = true does not log warnings

## Types of changes

- [ x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ x] Local run of `mvn install` succeeds
- [ x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x ] I have added tests to cover my changes
- [ x] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerFactory.java [hunk 1]
```diff
@@ -68,7 +64,7 @@ private static Supplier<S3AsyncClient> defaultS3AsyncClient() {
         if (crtInClasspath()) {
             return S3AsyncClient::crtCreate;
         }
-        return S3AsyncClient::create;
+        return () -> S3AsyncClient.builder().multipartEnabled(true).build();
```
Comment (davidh44) [inline services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerFactory.java:67]: Can we update javadocs in `S3TransferManager` to specify multipart enabled cases

File: services/s3/src/main/resources/codegen-resources/customization.config [hunk 2]
```diff
@@ -271,6 +270,10 @@
     "CrossRegionAccessEnabled":{
       "documentation":"Enables cross-region bucket access for this client",
       "type":"boolean"
+    },
+    "multipartEnabled":{
```
Comment (L-Applin) [inline services/s3/src/main/resources/codegen-resources/customization.config:274]: This also adds the multipart-enable config and method to the sync version of the client, as it adds the method the `S3BaseClientBuilder` interface which is used by both the sync and async builders:
```
S3Client client = S3Client.builder()
    .multipartEnabled(true) // <--- should not be available on sync builder
    .build();
```
I think we should avoid being able to define this config on the sync client and keep it on the async client only

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerFactory.java (hunk 1 above)
Comment (anirudh9391) [reply services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerFactory.java:67]: done

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerFactory.java [hunk 3]
```diff
@@ -52,11 +52,8 @@ public static S3TransferManager createTransferManager(DefaultBuilder tmBuilder)
             return new CrtS3TransferManager(transferConfiguration, s3AsyncClient, isDefaultS3AsyncClient);
         }
 
-        if (s3AsyncClient.getClass().getName().equals("software.amazon.awssdk.services.s3.DefaultS3AsyncClient")) {
-            log.warn(() -> "The provided DefaultS3AsyncClient is not an instance of S3CrtAsyncClient, and thus multipart"
-                           + " upload/download feature is not enabled and resumable file upload is not supported. To benefit "
-                           + "from maximum throughput, consider using S3AsyncClient.crtBuilder().build() instead.");
-        } else {
+        if (!s3AsyncClient.getClass().getName()
+                          .equals("software.amazon.awssdk.services.s3.internal.multipart.MultipartS3AsyncClient")) {
```
Comment (L-Applin) [inline services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerFactory.java:56]: How about also updation the log message to mention with something like:
> The provided S3AsyncClient is neither an instance of S3CrtAsyncClient or MultipartS3AsyncClient, and thus multipart upload/download feature may not be enabled and resumable file upload may not be supported.
To benefit from maximum throughput, consider either using S3AsyncClient.crtBuilder().build() or S3AsyncClient.builder().multipartEnabled(true).build() instead.

File: services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerLoggingTest.java
Comment (L-Applin) [inline services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerLoggingTest.java:1]: Could you add tests that also call `.crossRegionAccessEnabled(true)` on the s3 async client builder? The cross region builder method leads to creating `DelegatingS3AsyncClient` and I want to make sure the new logic still see the `MultipartS3AsyncClient` type.

Comment (anirudh9391) [reply services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerLoggingTest.java:1]: Done. The test passes. Publishing a new revision

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerFactory.java (hunk 3 above)
Comment (anirudh9391) [reply services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerFactory.java]: Changed.

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerFactory.java [hunk 4]
```diff
@@ -52,13 +52,12 @@ public static S3TransferManager createTransferManager(DefaultBuilder tmBuilder)
...
-                           + " upload/download feature is not enabled and resumable file upload is not supported. To benefit "
-                           + "from maximum throughput, consider using S3AsyncClient.crtBuilder().build() instead.");
-        } else {
-            log.debug(() -> "The provided S3AsyncClient is not an instance of S3CrtAsyncClient, and thus multipart"
-                            + " upload/download feature may not be enabled and resumable file upload may not be supported.");
+        if (!s3AsyncClient.getClass().getName().equals("software.amazon.awssdk.services.s3.internal.multipart"
+                                                       + ".MultipartS3AsyncClient")) {
+            log.debug(() -> "The provided S3AsyncClient is neither an instance of S3CrtAsyncClient or MultipartS3AsyncClient, " +
+                            "and thus multipart upload/download feature may not be enabled and resumable file upload may not " +
+                            "be supported.\n" + "To benefit from maximum throughput, consider using S3AsyncClient.crtBuilder().build() " +
```
Comment (L-Applin) [inline services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/TransferManagerFactory.java:59]: very nit-picky: I'd avoid \n in logs

**Comment by L-Applin** [comment]
There seems to still be a checkstyle violation:
```
2024-02-12T18:43:35.7470507Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.2:check (checkstyle) on project s3-transfer-manager: Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.42 with software/amazon/awssdk/checkstyle.xml ruleset. -> [Help 1]
```

**Review by L-Applin — APPROVED** [review]

**Pull request merged**